Frameworks/include/MockClass.h

```diff
@@ -0,0 +1,824 @@
+#include "WexTestClass.h"
+
+#include <functional>
+#include <wrl/implements.h>
+#include <wrl/client.h>
+
+/*/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+This header includes a number of macros that can be used to create mock classes for use in unit testing. The macros
+follow the format of MOCK_METHOD_n(functionName), where "n" is the number of parameters and "functionName" is the
+name of a function in one of the interfaces that the mock class implements.
+
+Internally these macros will implement the interface methods given as well as create a Set[functionName]
+setter per implemented function so that test code can set a custom C++ lambda that will be executed in place of
+the given interface function.
+
+Note: There are two sets of macros; one set for regular C++, and one set for high level C++/CX (see next section).
+
+This is all best explained with some sample code:
+
+struct Foo
+{
+int a;
+int b;
+};
+
+struct Bar
+{
+int a;
+int b;
+};
+
+class ISession
+{
+public:
+virtual void VoidNoArgs() = 0;
+virtual HRESULT AddMessageChannelListener(const Foo& foo) const = 0;
+virtual HRESULT RemoveMessageChannelListener(const Bar& bar, const Foo& foo) = 0;
+virtual HRESULT SendOnMessageChannel(const wchar_t* psz1, int x, char w, long long xyz) = 0;
+};
+
+class MockSessionWithLambdas : public ISession
+{
+public:
+MOCK_METHOD_0(VoidNoArgs);
+MOCK_CONST_METHOD_1(AddMessageChannelListener);
+MOCK_METHOD_2(RemoveMessageChannelListener);
+MOCK_METHOD_4(SendOnMessageChannel);
+
+MockSessionWithLambdas()
+{
+// Initialize callbacks
+SetVoidNoArgs([]() { });
+SetAddMessageChannelListener([](const Foo& foo){ return E_NOTIMPL; });
+SetRemoveMessageChannelListener([](const Bar& bar, const Foo& foo){ return E_NOTIMPL; });
+SetSendOnMessageChannel([](const wchar_t* psz1, int x, char w, long long xyz){ return E_NOTIMPL; });
+}
+};
+
+TEST_METHOD(InterfaceMockingSample)
+{
+MockSessionWithLambdas x;
+HRESULT result = x.AddMessageChannelListener(Foo());
+
+// In this example, result should now be "E_NOTIMPL" based on the default mock implementation of
+// AddMessageChannelListener.
+
+x.SetAddMessageChannelListenerCallback([](const Foo& foo)
+{
+VERIFY_ARE_EQUAL(32, foo.a); // Perform validation within the callback
+return HRESULT_FROM_WIN32(ERROR_OUT_OF_PAPER);
+});
+
+result = x.AddMessageChannelListener(Foo());
+
+// Now, result should be "HRESULT_FROM_WIN32(ERROR_OUT_OF_PAPER)" based on the lambda provided. In a real test
+// the MockSessionWithLambdas would passed to other classes that rely on ISession objects. Then, as those other
+// objects interact with the ISession interface, all of the provided lambdas will be executed, allowing
+// fine-grained control of how the other objects behave.
+
+VERIFY_ARE_EQUAL(HRESULT_FROM_WIN32(ERROR_OUT_OF_PAPER), result);
+}
+*/ /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```

Comment:
probably want to rewrite this example to use gtest syntax instead of TAEF syntax since we standardized on gtest for our test verification macros.